### DISCUSSION TITLE
How can I get all users currently in a channel?

### DISCUSSION THREAD

**Discussion by Jorge1701** [bwmarrin/discordgo#1233] [Q&A]
# Question

How can I get a list of users given a channel id?

I'm trying to make a command that shuffles the list of connected users in a voice channel.

But I'm having trouble getting the list.

# I have tried

Creating a new discord server.

Inviting my bot with all permissions.

Joining the General voice channel.

Running the following code, it recognizes the channel, but it gives me an empty array, like so:

```
General
[]
```

And if I use the commented code, since in some places it says that I should access some data throw the State struct, it gives me an error `state cache not found`

```go
package main

import (
	"fmt"

	"github.com/bwmarrin/discordgo"
)

func main() {
	session, err := discordgo.New(fmt.Sprintf("Bot %s", "YOUR_BOT_TOKEN"))

	if err != nil {
		fmt.Println("Error creating Discord session,", err)
		return
	}

	err = session.Open()
	if err != nil {
		fmt.Println("Error opening connection,", err)
		return
	}

	session.Identify.Intents = discordgo.IntentsAll

	//channel, err := session.State.Channel("YOUR_CHANNEL_ID")
	channel, err := session.Channel("YOUR_CHANNEL_ID")

	if err != nil {
		fmt.Println(err)
	} else {
		fmt.Println(channel.Name)
		fmt.Println(channel.Members)
	}

	session.Close()
}
```

**Comment by Nv7-GitHub** [comment]
You may want to wait for the `Ready` event so that the state will be populated, and also make sure you have given your bot the Guild Members privileged intent in the Discord Developer Portal.

**Comment by Jorge1701** [comment]
@Nv7-GitHub thanks for the response.

I've activated the privileged, now my Bot has that, and kicked the boy and made it join again.
![imagen](https://user-images.githubusercontent.com/26526255/186048090-86c77dbd-3951-4db3-ab24-0aeeafd9e52c.png)

And my code is the following now
```go
package main

import (
	"fmt"
	"os"
	"os/signal"
	"syscall"

	"github.com/bwmarrin/discordgo"
)

func ready(s *discordgo.Session, event *discordgo.Ready) {
	fmt.Println("Ready")

	//channel, err := session.State.Channel("YOUR_CHANNEL_ID")
	channel, err := s.Channel("YOUR_CHANNEL_ID")

	if err != nil {
		fmt.Println(err)
	} else {
		fmt.Println(channel.Name)
		fmt.Println(channel.Members)
	}
}

func main() {
	session, err := discordgo.New(fmt.Sprintf("Bot %s", "BOT_TOKEN"))

	if err != nil {
		fmt.Println("Error creating Discord session,", err)
		return
	}

	session.AddHandler(ready)

	session.Identify.Intents = discordgo.IntentsGuildMembers

	err = session.Open()
	if err != nil {
		fmt.Println("Error opening connection,", err)
		return
	}

	// Wait for closing signal
	sc := make(chan os.Signal, 1)
	signal.Notify(sc, syscall.SIGINT, syscall.SIGTERM, os.Interrupt, os.Kill)
	<-sc

	session.Close()
}
```
But I still get an empty slice when I'm in the `General` voice channel and run the code.
```
Ready
General
[]
```
I'm pretty sure it's some sort of permissions problem, but I can't figure out which or where to set it.

**Comment by Nv7-GitHub** [comment]
If you look at the [documentation](https://pkg.go.dev/github.com/bwmarrin/discordgo#Channel), it has
```go
// All thread members. State channels only.
Members []*ThreadMember `json:"-"`
```
Are you getting it from the state?

**Reply by Jorge1701** [reply]
What does getting it from the state means?

I have tried the this:
```
channel, err := session.State.Channel("YOUR_CHANNEL_ID")
```
Instead of
```
channel, err := session.Channel("YOUR_CHANNEL_ID")
```
But it gives me the following message: `state cache not found`

**Reply by FedorLap2006** [reply]
Sorry for a late reply, that just means the channel isn't in cache.

**Comment by koo04** [comment]
Because it was not clear, I did a lot of digging and finally answered this question. Check it out over [here](https://github.com/bwmarrin/discordgo/discussions/1273). Here is the example code I wrote up to see how it works.
```go
package main

import (
	"fmt"
	"time"

	"github.com/bwmarrin/discordgo"
)

func testCommandHandler(s *discordgo.Session, i *discordgo.InteractionCreate) {
	guild, err := s.State.Guild(i.GuildID)
	if err != nil {
		panic(err)
	}

	if err := s.InteractionRespond(i.Interaction, &discordgo.InteractionResponse{
		Type: discordgo.InteractionResponseChannelMessageWithSource,
		Data: &discordgo.InteractionResponseData{
			Flags:   discordgo.MessageFlagsEphemeral,
			Content: fmt.Sprintf("just a test: %d", len(guild.VoiceStates)),
		},
	}); err != nil {
		panic(err)
	}
}

func startBot() {
	fmt.Println("starting new discord bot")
	b, err := discordgo.New("Bot secret-token")
	if err != nil {
		panic(err)
	}

	b.Identify.Intents = discordgo.IntentsAll

	b.AddHandler(testCommandHandler)

	if err := b.Open(); err != nil {
		panic(err)
	}

	if _, err := b.ApplicationCommandCreate("app-id", "guild-id", &discordgo.ApplicationCommand{
		Name:        "test",
		Description: "Get the number of members in all voice channel",
	}); err != nil {
		panic(err)
	}
}

func main() {
	go startBot()

	for {
		fmt.Println("running")
		time.Sleep(10 * time.Second)
	}
}
```
